modules/system/assets/js/snowboard/ajax/handlers/AttributeRequest.js

```diff
@@ -0,0 +1,327 @@
+     */
+    keyDownHandler(event) {
+        // Check that we are inputting into a valid element
+        if (!event.target.matches(
+            'input',
+        )) {
+            return;
+        }
+
+        // Check that the input type is valid
+        const validTypes = [
+            'checkbox',
+            'color',
+            'date',
+            'datetime',
+            'datetime-local',
+            'email',
+            'image',
+            'month',
+            'number',
+            'password',
+            'radio',
+            'range',
+            'search',
+            'tel',
+            'text',
+            'time',
+            'url',
+            'week',
+        ];
+        if (validTypes.indexOf(event.target.getAttribute('type')) === -1) {
+            return;
+        }
+
+        if (event.key === 'Enter' && event.target.matches('*[data-request]')) {
+            this.processRequestOnElement(event.target);
+            event.preventDefault();
+            event.stopImmediatePropagation();
+        } else if (event.target.matches('*[data-track-input]')) {
+            this.trackInput(event.target);
+        }
+    }
+
+    /**
+     * Handles form submissions.
+     *
+     * @param {Event} event
+     */
+    submitHandler(event) {
+        // Check that we are submitting a valid form
+        if (!event.target.matches(
+            'form[data-request]',
+        )) {
+            return;
+        }
+
+        event.preventDefault();
+
+        this.processRequestOnElement(event.target);
+    }
+
+    /**
+     * Processes a request on a given element, using its data attributes.
+     *
+     * @param {HTMLElement} element
+     */
+    processRequestOnElement(element) {
+        const data = element.dataset;
+
+        const handler = String(data.request);
+        const options = {
+            confirm: ('requestConfirm' in data) ? String(data.requestConfirm) : null,
+            redirect: ('requestRedirect' in data) ? String(data.requestRedirect) : null,
+            loading: ('requestLoading' in data) ? String(data.requestLoading) : null,
+            flash: ('requestFlash' in data),
+            files: ('requestFiles' in data),
+            browserValidate: ('requestBrowserValidate' in data),
+            form: ('requestForm' in data) ? String(data.requestForm) : null,
+            url: ('requestUrl' in data) ? String(data.requestUrl) : null,
+            update: ('requestUpdate' in data) ? this.parseData(String(data.requestUpdate)) : [],
+            data: ('requestData' in data) ? this.parseData(String(data.requestData)) : [],
+        };
+
+        this.snowboard.request(element, handler, options);
+    }
+
+    /**
+     * Sets up an AJAX request via HTML attributes.
+     *
+     * @param {Request} request
+     */
+    onAjaxSetup(request) {
+        const fieldName = request.element.getAttribute('name');
+
+        const data = {
+            ...this.getParentRequestData(request.element),
```

Comment:
Is this just the function that allows a data-request on a button element to include data-request-data defined on the form element above? Or is it related to the PR that I linked earlier? Also does it mirror the original behaviour, I could have sworn that the data only included up to the first form element?

Reply:
Yep, exactly. It includes the data from any parent element in the final "compiled" data.

The previous behaviour was here: https://github.com/wintercms/winter/blob/develop/modules/system/assets/js/framework.js#L94-L96

As far as I can tell, it got the data from any element - it didn't stop at the nearest form.